### ISSUE TITLE
Conv2D feeding into LSTM breaks model for inference

### ISSUE THREAD

**Issue by stefan-falk** [tensorflow/tensorflow#47034]
**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 18.04
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): 2.4.0
- Python version: 3.8

**Describe the current behavior**

Having a `Conv2D` layer with a `filter_size=(1, F)`, where `F` is the number of features, delivers inconsistent results while feeding a chunked stream to a model.

I identified this `Conv2D` layer as the breaking-point in my model for speech-recognition. The model works fine _without_ the `Conv2D` layer using chunking and non-chunking. However, if `Conv2D` is present, chunking is not working anymore.

**Describe the expected behavior**

The error between the chunked and non-chunked model output should _always_ be exactly `0.0` or at least *very* close.

**Standalone code to reproduce the issue**

The code below generates plot, as shown below, to illustrate the (potential) issue.

Essentially, it generates a toy-model which feeds inputs, taking the form `(batch, time, features, channels)` into the `Conv2D` layer and project this down to `(batch, time, 1, filters)`. Note the `tf.concat` before the `Conv2D` layer which artificially creates this input in the example. 

After the convolution we just drop the third axis and feed the result into the LSTM layers.

```python
import string

import numpy as np
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import layers


def get_model(
    vocab_size: int,
    embedding_size: int,
    hidden_size: int,
    conv2d_kwargs: dict = None
):
    inputs = layers.Input(shape=(None,), dtype=tf.int32)

    embedding = layers.Embedding(
        vocab_size + 1,
        embedding_size,
        mask_zero=True
    )
    x = embedding(inputs)

    if conv2d_kwargs is not None:
        x = layers.Lambda(lambda l: tf.expand_dims(l, axis=-1))(x)
        x = layers.Lambda(lambda t: tf.concat([t, t, t], axis=-1))(x)
        x = layers.Conv2D(**conv2d_kwargs)(x)
        x = layers.Lambda(lambda l: tf.squeeze(l, axis=2))(x)

    lstm_out = layers.LSTM(hidden_size, return_sequences=True, return_state=True)(x)
    x = lstm_out[0]
    lstm_out = layers.LSTM(hidden_size, return_sequences=False, return_state=True)(x)
    x = lstm_out[0]
    outputs = layers.Dense(vocab_size, activation='softmax')(x)
    return keras.Model(inputs=inputs, outputs=outputs)


def infer(model, inputs, state=None):
    x = inputs
    new_states = list()

    # We're just interested in the output of the Conv2D layer and the succeeding LSTM-layer
    layers_of_interest = model.layers[:-2]

    for layer in layers_of_interest:
        if isinstance(layer, layers.LSTM):
            idx = len(new_states)
            outputs = layer(x, initial_state=state[idx] if state is not None else None)
            x, new_state = outputs[0], outputs[1:]
            new_states.append(new_state)
        else:
            x = layer(x)

    return x, new_states


def eval_plot(model, sample_text: str, char2idx: dict, chunk_size: int):

    enc_chunks = list()
    state_trained = None
    nb_chunks = int(np.ceil(len(sample_text) / chunk_size))
    for i in range(nb_chunks):
        s = i * chunk_size
        e = s + chunk_size
        text_chunk = sample_text[s:e]
        test_inputs = list(map(lambda c: char2idx[c], text_chunk))
        test_inputs = tf.constant([test_inputs], dtype=tf.int32)
        encoded, state_trained = infer(model, test_inputs, state=state_trained)
        enc_chunks.append(encoded)

    test_inputs = list(map(lambda c: char2idx[c], sample_text))
    test_inputs = tf.constant([test_inputs], dtype=tf.int32)
    enc_full, _ = infer(model, test_inputs)
    enc_full = enc_full.numpy()[0]

    chunks_concat = tf.concat(enc_chunks, axis=1).numpy()[0]
    diff = chunks_concat - enc_full
    rmspe = (np.sqrt(np.mean(np.square(diff / enc_full)))) * 100
    import matplotlib.pyplot as plt
    fix, axes = plt.subplots(1, 3, figsize=(16, 9), sharex='all', sharey='all')
    ax1, ax2, ax3 = axes
    ax1.matshow(chunks_concat.T)
    ax1.set_title('Chunked')
    ax2.matshow(enc_full.T)
    ax2.set_title('Full')
    ax3.matshow(diff.T)
    ax3.set_title(f'Diff (RMSPE: {rmspe:.8f})')
    for ax in axes:
        ax.set_xlabel('Time [T]')
    ax1.set_ylabel('Features [F]')
    plt.tight_layout()
    plt.show()
    return rmspe


def main():

    vocab = set(' ' + string.ascii_lowercase)
    vocab_size = len(vocab)
    idx2char = {(idx + 1): char for idx, char in enumerate(vocab)}
    char2idx = {char: idx for idx, char in idx2char.items()}

    units = 32
    embedding_size = 16
    filters = units * 2

    model_conv_params = dict(
        vocab_size=vocab_size,
        embedding_size=embedding_size,
        hidden_size=units,
        conv2d_kwargs=dict(filters=filters, kernel_size=(1, embedding_size), use_bias=False)
    )

    model_conv = get_model(**model_conv_params)

    sample_text = 'how are you buddy'

    for chunk_size in range(1, 11):
        print(f'Creating plots for chunk size {chunk_size}')
        rmspe = eval_plot(model_conv, sample_text, char2idx, chunk_size)
        print(f'  RMSPE: {rmspe:.4f}')

    print('All done')


if __name__ == '__main__':
    main()

```

### Toy-example plots

![image](https://user-images.githubusercontent.com/43335432/107357656-a93dad00-6ad2-11eb-9e15-5ce79c3a7f21.png)

![image](https://user-images.githubusercontent.com/43335432/107357740-bfe40400-6ad2-11eb-9897-a350652e3ba4.png)

### "Real" Model Plots

The following are plots from a speech recognition model which uses this simple `Conv2D` operation before it is send into LSTM layers. One plot is after 1000 steps of training, the other after 200. As you can see, **the MSE, computed for each frame in the last image, increases over time** (note the y-axis range). This model is "broken" **using chunking** and does not produce any meaningful output anymore after training it until early stopping. It does work *without* chunking though.

![image](https://user-images.githubusercontent.com/43335432/107615058-c9dd4280-6c4b-11eb-9b7b-2cf8169cfbf2.png)

![image](https://user-images.githubusercontent.com/43335432/107615064-ce096000-6c4b-11eb-8da7-519df0e33219.png)

On the other hand, using the same model but removing the `Conv2D` layer gives an MSE for each frame somewhere below `6e-14` which is (basically) zero:

![image](https://user-images.githubusercontent.com/43335432/107615288-293b5280-6c4c-11eb-837e-857c6773efb2.png)

**Comment by stefan-falk** [comment]
Here is another example where I use a `layers.Dense` instead of `Conv2D`. This is _not_ a replacement for my use-case but it's weird that this shows the exact same behavior as the example I posted above:

```python
import tensorflow as tf
import numpy as np
import matplotlib.pyplot as plt

from tensorflow.keras import layers
from tensorflow import keras

def infer(model, inputs, state=None):
    x = inputs
    new_states = list()

    for layer in model.layers:
        if isinstance(layer, layers.LSTM):
            idx = len(new_states)
            outputs = layer(x, initial_state=state[idx] if state is not None else None)
            x, new_state = outputs[0], outputs[1:]
            new_states.append(new_state)
        else:
            x = layer(x)
    return x, new_states


def reshape_(t):
    shape = tf.shape(t)
    B, T, F, C = t.get_shape().as_list()
    t = tf.reshape(t, shape=(shape[0], shape[1], shape[2] * shape[3]))
    t.set_shape((B, T, F * C))
    return t


rnd = np.random.RandomState(42)
feat_shape = (50, 66, 3)
batch = rnd.rand(*((1,) + feat_shape))

# Build the model

filters = 16
lstm_units = 12
F = feat_shape[-2]

inputs = layers.Input(shape=(None,) + feat_shape[1:], dtype=tf.float32)
x = inputs
x = layers.Lambda(reshape_)(x)
x = layers.Dense(filters)(x)
lstm_out = layers.LSTM(lstm_units, return_sequences=True, return_state=True)(x)
x = lstm_out[0]

model = keras.Model(inputs=inputs, outputs=x)

# Reference

enc = model(batch)[0].numpy()

# Chunk-wise inference

for chunk_size in [1, 2, 3, 4, 5, 6]:
    nb_chunks = int(np.ceil(batch.shape[1] / chunk_size))

    state = None
    enc_chunks = list()
    for i in range(nb_chunks):
        s = i * chunk_size
        e = s + chunk_size
        chunk = batch[:, s:e]
        enc_chunk, state = infer(model, chunk, state=state)
        enc_chunks.append(enc_chunk)

    enc_chunk_concat = tf.concat(enc_chunks, axis=1)[0].numpy()
    
    diff = (enc_chunk_concat - enc)
    rmspe = np.sqrt(np.mean((diff / enc)**2))

    fig, axes = plt.subplots(3, 1, figsize=(16, 9))
    ax1, ax2, ax3 = axes
    ax1.matshow(enc_chunk_concat.T)
    ax1.set_title('Chunked')
    ax2.matshow(enc.T)
    ax1.set_title('Reference')
    ax3.matshow(diff.T)
    ax1.set_title(f'Diff')
    fig.suptitle(f'chunk_size={chunk_size}, RMSPE: {rmspe:.8f}%')
    plt.show()
```

**Comment by Saduf2019** [comment]
@stefan-falk 
I ran the code shared and face  a different [issue](https://colab.research.google.com/gist/Saduf2019/dafb8e5520c958cb8dfb5473f82357aa/untitled529.ipynb), please share all dependencies for us to replicate the issues.

**Comment by stefan-falk** [comment]
@Saduf2019 The code in my first post is a standalone and should run. Why do you have a call to `image_dataset_from_directory()`?

However, the interesting thing I noticed is that running the code in Colab does _not_ show the same results..

**Comment by Saduf2019** [comment]
@ymodak
I ran the code on tf 2.4 and nightly, please find the [gist here]( 
https://colab.research.google.com/gist/Saduf2019/6bf97a0ebe14110f15813eefe6161856/untitled529.ipynb)

**Comment by stefan-falk** [comment]
fyi: I was able to fix my original problem. In turns out that I had a bug elsewhere which broke my model. For this matter, this issue is not relevant anymore.

However, it would be interesting to know why the example above sometimes produces sometimes an error of 0 and sometime one that's > 0.

It's very likely that these are just some rounding/floating errors but it's kind of interesting to see that there's no difference at all in at least some cases.

**Comment by ymodak** [comment]
Closing this issue since the original query is resolved. Thanks!
